### ISSUE TITLE
mw/proxy

### ISSUE THREAD

**Issue by UladzimirTrehubenka** [coredns/coredns#1035]
Faced with the issue:
on the project we use CoreDNS configured for using Unbound resolver as a caching DNS proxy.
If Unbound sends to external resolver some query and it responses with RcodeServerFailure (Unbound returns this answer to CoreDNS), CoreDNS makes decision "unreachable backend" (Unbound) and some queries in interval fail_timeout (default 10 sec) are also resolved as RcodeServerFailure (i.e. ignore proxy backend as down instance). As workaround we just set max_fails to 0 to avoid this.

Would be useful to implement enhancement - add option:
dns_health_check DURATION dns_name1, dns_name2, dns_name3 ...

On the project we use following list: "amazon.com", "google.com", "youtube.com", "facebook.com", "yahoo.com", "apple.com", "microsoft.com" (suppose that these names should be resolved always). The logic is: get first name and make sure that query ret code is RcodeSuccess, if not - get next name until RcodeSuccess or end of the list. This dns_health_check allows determine that proxy endpoint works (queries are resolved) and network route between proxy endpoint and the Internet is fine (option health_check not so useful - looks like inherited from caddy http feature).

**Comment by miekg** [comment]
I think I know what you're hinting at, but RcodeServerFailure shouldn't trigger this - but I should double check

**Comment by nrvnrvn** [comment]
I've hit the same issue.
Getting a bunch of errors like this:

```
[13/Sep/2017:09:57:13 +0000] "A IN notifications.google.com. udp 42 false 512" SERVFAIL 42 171.123µs
13/Sep/2017:09:57:13 +0000 [ERROR 0 notifications.google.com. A] unreachable backend
```

Less than a millisecond duration and eyeballing the tcpdumps show that coredns does not even make calls to the configured dns (google public).

Adding `max_fails 0` seems to really solve the issue.

Version: 010
Config: 
```
.:53 {
    proxy . 8.8.8.8:53 8.8.4.4:53 {
        max_fails 0
    }
    log {
        class error
    }
    errors
}
```

**Comment by miekg** [comment]
Commenting here for further idea on making this more robust (also see #486). 

<strike>Should we also implement retries?</strike> multiple upstream are tried in succession. This probably means that using `proxy . /etc/resolv.conf` is worse than `proxy . 169.254.169.254 169.254.169.254` in GCE for instance.

**Comment by miekg** [comment]
An idea raised in #486 (or somewhere else) by @johnbelamaric is to use a `. IN NS` query for health checking. I like that; it's a bit nonsensical to then also have http health checks?

**Comment by miekg** [comment]
See https://github.com/coredns/coredns/pull/1093 for a possible solution.

**Comment by miekg** [comment]
Another question while working on this. Making DNS health checks in band makes sense, but my current approach replaces the HTTP health checks completely. Is that something that is OK, or should we keep those as well.

The new apiproxy healthchecks in the kubernetes plugin utilize the HTTP healtchecks.

**Comment by miekg** [comment]
new plan. Rename plugin/pkg/health to plugin/pkg/httphealth convert current users (kubernetes). Reimplement healtchecks in the proxy package (as before) and make it in band

**Comment by johnbelamaric** [comment]
Makes sense.

**Comment by miekg** [comment]
Ok, the more I code the larger the PR will get and it is not feeling right (`dns-ping2` is the branch). Also doing *out of band* healthchecks makes and with http return code we can even do lame-ducking to drain upstreams.

I'm going to push some cleans that I spotted in separate PRs and then fix this in another way.

**Comment by miekg** [comment]
New, new plan after https://github.com/coredns/coredns/pull/1106 is merged.

Handle `Timeout() == true` errors differently and kick off a healthcheck if fails>3

**Comment by miekg** [comment]
Can either of you try the latest master? I can provide binaries if needed.

@johnbelamaric I'm not comfortable kicking off healthchecks on upstream failures; the healthchecks are tcp/http meaning you can potentially get a lot of these from someone fiddling with UDP stream.

**Comment by johnbelamaric** [comment]
@UladzimirTrehubenka if this is fixed by #1110 please close this issue.

**Comment by miekg** [comment]
Closing because of staleness
